lib/trino-filesystem-s3/src/main/java/io/trino/filesystem/s3/S3OutputStream.java

```diff
@@ -0,0 +1,342 @@
+import software.amazon.awssdk.services.s3.model.CompletedPart;
+import software.amazon.awssdk.services.s3.model.CreateMultipartUploadRequest;
+import software.amazon.awssdk.services.s3.model.PutObjectRequest;
+import software.amazon.awssdk.services.s3.model.RequestPayer;
+import software.amazon.awssdk.services.s3.model.UploadPartRequest;
+
+import java.io.IOException;
+import java.io.InterruptedIOException;
+import java.io.OutputStream;
+import java.nio.ByteBuffer;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Optional;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.Future;
+
+import static com.google.common.primitives.Ints.constrainToRange;
+import static java.lang.Math.max;
+import static java.lang.Math.min;
+import static java.lang.System.arraycopy;
+import static java.util.Objects.requireNonNull;
+import static java.util.concurrent.CompletableFuture.supplyAsync;
+import static software.amazon.awssdk.services.s3.model.ServerSideEncryption.AES256;
+import static software.amazon.awssdk.services.s3.model.ServerSideEncryption.AWS_KMS;
+
+final class S3OutputStream
+        extends OutputStream
+{
+    private final byte[] oneByte = new byte[1];
+    private final List<CompletedPart> parts = new ArrayList<>();
+    private final LocalMemoryContext memoryContext;
+    private final S3Client client;
+    private final S3Location location;
+    private final int partSize;
+    private final RequestPayer requestPayer;
+    private final S3SseType sseType;
+    private final String sseKmsKeyId;
+
+    private int currentPartNumber;
+    private byte[] buffer = new byte[0];
+    private int bufferSize;
+    private int initialBufferSize = 64;
+
+    private boolean closed;
+    private boolean failed;
+    private boolean multipartUploadStarted;
+    private Future<CompletedPart> inProgressUploadFuture;
+
+    // Mutated by background thread which does the multipart upload.
+    // Read by both main thread and background thread.
+    // Visibility is ensured by calling get() on inProgressUploadFuture.
+    private Optional<String> uploadId = Optional.empty();
+
+    public S3OutputStream(AggregatedMemoryContext memoryContext, S3Client client, S3Context context, S3Location location)
+    {
+        this.memoryContext = memoryContext.newLocalMemoryContext(S3OutputStream.class.getSimpleName());
+        this.client = requireNonNull(client, "client is null");
+        this.location = requireNonNull(location, "location is null");
+        this.partSize = context.partSize();
+        this.requestPayer = context.requestPayer();
+        this.sseType = context.sseType();
+        this.sseKmsKeyId = context.sseKmsKeyId();
+    }
+
+    @SuppressWarnings("NumericCastThatLosesPrecision")
+    @Override
+    public void write(int b)
+            throws IOException
+    {
+        oneByte[0] = (byte) b;
+        write(oneByte);
```

Comment:
consider inlining this directly and removing the `onebyte` field.... I believe this inlines to like 3-4 lines of code